### ISSUE TITLE
Sync not checking for library data changes

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero-ios#381]
I created a couple items on the desktop and synced:

```
[11/Nov/2021:04:26:01 -0500] "GET /keys/current HTTP/1.1" 200 176 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
[11/Nov/2021:04:26:01 -0500] "GET /users/12345/groups?format=versions HTTP/1.1" 200 2 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
[11/Nov/2021:04:26:01 -0500] "POST /users/12345/items HTTP/1.1" 200 1401 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
[11/Nov/2021:04:26:01 -0500] "POST /users/12345/items HTTP/1.1" 200 584 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
[11/Nov/2021:04:26:01 -0500] "GET /users/12345/fulltext?format=versions&since=1220 HTTP/1.1" 200 2 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
[11/Nov/2021:04:26:01 -0500] "POST /users/12345/fulltext HTTP/1.1" 200 99 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:60.0) Gecko/20100101 Zotero/5.0.97-beta.53+f583416d8"
```

The streaming server sent notifications to both clients:

```
[11/Nov/2021:04:26:05 -0500] Send: {"event":"topicUpdated","topic":"/users/12345","version":1223}
[11/Nov/2021:04:26:05 -0500] Send: {"event":"topicUpdated","topic":"/users/12345","version":1223}
```

The iOS app then synced, but it didn't actually check for data changes, and pulling down on the items list didn't cause it to:

```
[11/Nov/2021:04:26:08 -0500] "GET /keys/current HTTP/1.1" 200 176 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:26:23 -0500] "GET /keys/current HTTP/1.1" 200 176 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:26:23 -0500] "GET /users/12345/groups?format=versions HTTP/1.1" 200 2 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:26:56 -0500] "GET /keys/current HTTP/1.1" 200 176 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:26:56 -0500] "GET /users/12345/groups?format=versions HTTP/1.1" 200 2 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:27:31 -0500] "GET /keys/current HTTP/1.1" 200 176 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
[11/Nov/2021:04:27:31 -0500] "GET /users/12345/groups?format=versions HTTP/1.1" 200 2 "-" "Zotero/163 CFNetwork/1240.0.4 Darwin/20.6.0"
```

**Comment by michalrentka** [comment]
@dstillman if reproducible, could you also provide logs from iOS? The repeating of keys and group versions requests would indicate that the sync was aborted (and retried) after group version sync. Abort in group versions sync happens only if the request fails, but as we can see, it returned 200. So I'm not sure what's happening only based on those logs.

**Comment by dstillman** [comment]
D1450316290

It's not repeating — those were just separate manual sync attempts. Every time I pull down on the list, I got those two requests, and nothing else. Looks like there's a WebDAV error:

```
(+0000000): UploadAttachmentSyncAction: upload to WebDAV

(+0000000): UploadAttachmentSyncAction: check whether attachment item has been submitted

(+0000001): UploadAttachmentSyncAction: validate file to upload
[ERROR] Zotero(+0000001): UploadAttachmentSyncAction: missing attachment - file:///private/var/mobile/Containers/Shared/AppGroup/DCD5DAD0-E77B-4176-BE1C-518B1F2C44EC/downloads/custom_my_library/GY67ZIN6/a.txt
[ERROR] Zotero(+0000001): WebDavController: finish failed upload - attachmentMissing(key: "GY67ZIN6", title: "a")
[ERROR] Zotero(+0000000): UploadAttachmentSyncAction: could not upload - attachmentMissing(key: "GY67ZIN6", title: "a")
```

I realize now that it's actually reporting that in the bottom left. (I had ignored it because I always get one in another account where I have incompatible data.)

**Comment by dstillman** [comment]
That shouldn't interfere with data sync, though, which should happen first.

**Comment by dstillman** [comment]
A few things:

1) Seems like you're counting the attachment item part of a file uploads as a library action, such that a failure there would block data syncing? That's not how we handle things on the desktop — all parts of a file sync, including the initial attachment item write, happen strictly after data syncing. So you'd never have an item not appear just because there was a file sync error.

2) Why is it trying to upload a file that doesn't exist locally?

3) Weirdly, this file does actually exist on the WebDAV server, and when I try to open it I see a 200 for the ZIP. But it still shows the error icon, and then says the attached file could not be found.

**Comment by michalrentka** [comment]
> 1. Seems like you're counting the attachment item part of a file upload as a library action, such that a failure there would block data syncing? That's not how we want to do it — all parts of a file sync, including the initial attachment item write, should happen strictly after data syncing. So you'd never have an item not appear just because there was a file sync error.

Yes and no. I didn't separate between "item actions" and "file actions". I load changed items from DB (which also includes newly created attachments) and try to submit them. If a file is missing I get a non-fatal error, which doesn't block further sync (the error is just collected and shown at the end).

I can of course move the attachment-related actions to the end, but I don't think it would change much (except maybe item changes would be synced faster, since files would start uploading only after everything else was synced).

I actually realised the problem now. Normally, when a sync starts, there are either local changes up for submission or the sync tries to fetch remote changes. If there are local changes, it tries to submit them and the server can return 412, which indicates, that there are remote changes which need to be fetched first. 
But in your case, there is only a file upload remaining. So it tries to upload it. It can't even find a file locally. It returns a non-fatal error and tries to continue the queue of actions. But there are no more actions. Only this upload was enqueued. And since no http request happened because of local issue, we don't know that there might be remote changes.

I could check whether any HTTP requests for given library were performed before all library queued actions finished and if no request was performed, we should fetch remote changes. Of course ideally the file would not be missing, but bugs can happen in future in other places and we could end up stuck here again.

> 2. Why is it trying to upload a file that doesn't exist locally?

It sees in the database that a file is up for upload, so it tries to upload it, file is not found and non-fatal error is reported.

> 3. Weirdly, this file does actually exist on the WebDAV server, and when I try to open it I see a 200 for the ZIP. But it still shows the error icon, and then says the attached file could not be found.

That's weird. Sounds like the file was downloaded but couldn't be unzipped. What did the error say after tap? Do you have log from this?

**Comment by dstillman** [comment]
> I could check whether any HTTP requests for given library were performed before all library queued actions finished and if no request was performed, we should fetch remote changes.

Yup, though I wouldn't think of it as "HTTP requests" — e.g., a PUT request to WebDAV doesn't count. It'd have to be an API write that interacts with the library version and either goes through or gets a 412. (I assume this is what you meant.)

> What did the error say after tap? Do you have log from this?

Error is "The attachmed file could not be found." D1241484712

```
“a.txt” couldn’t be moved to “GY67ZIN6” because an item with the same name already exists.
```

**Comment by michalrentka** [comment]
@dstillman oh you actually have a .txt file on webdav? I thought it's supposed to have only zips. I always try to unzip in case of webdav.

**Comment by dstillman** [comment]
No, I have a ZIP containing a.txt.
